src/gfx/convert.cpp

```diff
@@ -0,0 +1,1001 @@
+                              std::vector<Palette> const &palettes,
+                              DefaultInitVec<size_t> const &mappings) {
+	// Iterate throughout the image, generating tile data as we go
+	// (We don't need the full tile data to be able to dedup tiles, but we don't lose anything
+	// by caching the full tile data anyway, so we might as well.)
+	UniqueTiles tiles;
+
+	auto iter = attrmap.begin();
+	for (auto tile : png.visitAsTiles(options.columnMajor)) {
+		auto [tileID, matchType] = tiles.addTile(tile, palettes[mappings[iter->protoPaletteID]]);
+
+		iter->xFlip = matchType == TileData::HFLIP || matchType == TileData::VHFLIP;
+		iter->yFlip = matchType == TileData::VFLIP || matchType == TileData::VHFLIP;
+		iter->bank = tileID >= options.maxNbTiles[0];
+		iter->tileID = (iter->bank ? tileID - options.maxNbTiles[0] : tileID)
+		               + options.baseTileIDs[iter->bank];
+
+		++iter;
+	}
+	assert(iter == attrmap.end());
+
+	// Copy elision should prevent the contained `unordered_set` from being re-constructed
+	return tiles;
+}
+
+static void outputTileData(UniqueTiles const &tiles) {
+	std::filebuf output;
+	output.open(options.output, std::ios_base::out | std::ios_base::binary);
+
+	uint16_t tileID = 0;
+	for (auto iter = tiles.begin(), end = tiles.end() - options.trim; iter != end; ++iter) {
+		TileData const *tile = *iter;
+		assert(tile->tileID == tileID);
+		++tileID;
+		output.sputn(reinterpret_cast<char const *>(tile->data().data()), options.bitDepth * 8);
+	}
+}
+
+static void outputTilemap(DefaultInitVec<AttrmapEntry> const &attrmap) {
+	std::filebuf output;
+	output.open(options.tilemap, std::ios_base::out | std::ios_base::binary);
+
+	for (AttrmapEntry const &entry : attrmap) {
+		output.sputc(entry.tileID); // The tile ID has already been converted
+	}
+}
+
+static void outputAttrmap(DefaultInitVec<AttrmapEntry> const &attrmap,
+                          DefaultInitVec<size_t> const &mappings) {
+	std::filebuf output;
+	output.open(options.attrmap, std::ios_base::out | std::ios_base::binary);
+
+	for (AttrmapEntry const &entry : attrmap) {
+		uint8_t attr = entry.xFlip << 5 | entry.yFlip << 6;
+		attr |= entry.bank << 3;
+		attr |= mappings[entry.protoPaletteID] & 7;
+		output.sputc(attr);
+	}
+}
+
+} // namespace optimized
+
+void process() {
+	options.verbosePrint(Options::VERB_CFG, "Using libpng %s\n", png_get_libpng_ver(nullptr));
+
+	options.verbosePrint(Options::VERB_LOG_ACT, "Reading tiles...\n");
+	Png png(options.input);
+	ImagePalette const &colors = png.getColors();
+
+	// Now, we have all the image's colors in `colors`
+	// The next step is to order the palette
+
+	if (options.verbosity >= Options::VERB_INTERM) {
+		fputs("Image colors: [ ", stderr);
+		size_t i = 0;
+		for (auto const &slot : colors) {
+			if (!slot.has_value()) {
+				continue;
+			}
+			fprintf(stderr, "#%02x%02x%02x%02x%s", slot->red, slot->green, slot->blue, slot->alpha,
```

Comment:
Can `slot` have a `toCss` method as well?

Reply:
Ah, but, you know, `slot` *has* a `toCSS` method already—it's a `std::optional<Rgba>`! xD